File: algorithms/0697.degree_of_an_array/degree_of_an_array.rs
use std::collections::HashMap;

impl Solution {
    pub fn find_shortest_sub_array(nums: Vec<i32>) -> i32 {
        let mut nums_degree = HashMap::new();
        for &n in nums.iter() {
            let count = nums_degree.entry(n).or_insert(0);
            *count += 1;
        }

        let degree = *nums_degree.values().max().unwrap();
        let mut degree_pos: HashMap<i32, Vec<usize>> = HashMap::new();
        for i in 0..nums.len() {
            let n = nums[i];
            if nums_degree.contains_key(&n) && nums_degree[&n] == degree {
                let indexs = degree_pos
                    .entry(n)
                    .or_insert(vec![]);
                indexs.push(i);
            }
        }
        degree_pos
            .values()
            .map(|v| v.last().unwrap() - v.first().unwrap())
            .min()
            .unwrap() as i32
            + 1
    }
}

File: algorithms/0697.degree_of_an_array/degree_of_an_array.cpp
class Solution {
public:
    int findShortestSubArray(const vector<int>& nums) {
        unordered_map<int, int> degrees;
        unordered_map<int, vector<size_t>> nums_index;
        int ans = INT_MAX;
        int max_counter = 0;
        for (size_t i = 0; i < nums.size(); ++i) {
            const int &n = nums[i];
            degrees[n] += 1;
            max_counter = max(max_counter, degrees[n]);
            nums_index[n].push_back(i);
        }
        for (const auto &[key, count]: degrees) {
            if (count == max_counter) {
                const int len = nums_index[key].back() - nums_index[key].front() + 1;
                ans = min(len, ans);
            }
        }
        return ans;
    }
};